react-custom-hooks/client/src/components/useTodos.js

```diff
@@ -0,0 +1,133 @@
+import { useEffect, useState } from 'react';
+
+const url = (path) => `${process.env.REACT_APP_BASE_URL}${path}`;
+
+/**
+ * Manages the Todos by reading from and writing to the backend API.
+ *
+ * @returns {Object} with the following properties:
+ *   `isLoading`: a boolean indicating if the data has not yet loaded;
+ *   `todos`: a list of To Do items;
+ *   `addTodo`: a function (newTodo) that will add a new To Do;
+ *   `toggleCompleted`: a function (todoId) that will toggle the completion of an existing To Do.
+ */
+export function useTodos() {
+  const [todos, setTodos] = useState();
+  useEffect(() => {
+    /* TODO: If `todos` hasn't been defined yet, read the items from the API
+     * and set them into the `todos` state.
+     */
+    async function fetchData() {
+      if (!todos) {
+        try {
+          const todosData = await readTodos();
+          setTodos(todosData);
+        } catch (error) {
+          console.error(error);
+        }
+      }
+    }
+    fetchData();
+  }, [todos]);
+
+  async function addTodo(newTodo) {
+    /* TODO: Call the API function that creates a Todo item.
+     * When the promise returned by that function resolves, update the `todos` state.
+     * Note that it is critical that you pass a _new_ array. Do not modify the `todos` array.
+     */
+    try {
+      const createdTodo = await createTodo(newTodo);
+      setTodos(prevTodos => [...prevTodos, createdTodo]);
+    } catch (error) {
+      console.error(error);
+    }
+  }
+
+  async function toggleCompleted(todoId) {
+    /* TODO: Find the Todo item being updated, toggle its completed prop, and call
+     * the API function that updates a Todo item.
+     * When the promise returned by that function resolves, update the `todos` state.
+     * When updating this state, use the updated `todo` returned from the API.
+     * Note that it is critical that you pass a _new_ array. Do not modify the `todos` array.
+     */
+    setTodos(prevTodos => {
```

Comment:
This makes the function rather long and difficult to read. It would be simpler to just do all the work, then call `setTodos` at the end.